File: dice/qtcpp/ini/apigear_game/factory.cpp
#include "factory.h"
#include "dice.h"
#include "random.h"

AbstractFactory *Factory::s_instance = nullptr;

Factory::Factory(QObject *parent)
    : AbstractFactory(parent)
{
}

void Factory::set(AbstractFactory *instance)
{
    if(s_instance) {
        qFatal("Can not set factory when factory already set");
    }
    s_instance = instance;
}

AbstractFactory * Factory::get()
{
    if(s_instance) {
        return s_instance;
    }
    s_instance = new Factory(QCoreApplication::instance());
    return s_instance;
}

AbstractDice* Factory::createDice(QObject *parent)
{
    return new Dice(parent);
}

AbstractRandom* Factory::createRandom(QObject *parent)
{
    return new Random(parent);
}

File: dice/typescript/ini/apigear_game/factory.ts
import {  
  IDice,  
  IRandom
} from './api'

import { Dice } from './dice'

import { Random } from './random'

export class Factory {  
  createDice() : IDice {
    return new Dice()
  }  
  createRandom() : IRandom {
    return new Random()
  }
}